### PR TITLE
firstboot: various enhancements

### PR DESCRIPTION
**Pull request by NekkoDroid**
1. Implement "default" values
	- When nothing is entered a compile time default is selected
	- Passing `-` is now used to skip the prompt
1. Some cleanups

See #31832

### CODE REVIEW

File: src/firstboot/firstboot.c [hunk 1]
```diff
@@ -97,7 +97,7 @@ static bool press_any_key(void) {
         char k = 0;
         bool need_nl = true;
 
-        printf("-- Press any key to proceed --");
+        printf("-- Press any key to proceed --\n");
         fflush(stdout);
```
Comment (poettering) [inline src/firstboot/firstboot.c]: you can drop the fflush() then

File: src/firstboot/firstboot.c [hunk 2]
```diff
@@ -210,7 +210,7 @@ static int prompt_loop(const char *text, char **l, unsigned percentage, bool (*i
                         return log_error_errno(r, "Failed to query user: %m");
 
                 if (isempty(p)) {
-                        log_warning("No data entered, skipping.");
+                        log_debug("No data entered, skipping.");
```
Comment (poettering) [inline src/firstboot/firstboot.c]: what's the rationale for this change?

Comment (NekkoDroid) [reply src/firstboot/firstboot.c]: @keszybz mentioned this in https://github.com/systemd/systemd/pull/31154#issuecomment-1946056086. The specific words used where 

> Stop saying "No data entered, skipping." at warning level. In particular in interactive mode this is not useful, the user knows that they pressed Enter.

Which I do kind of agree with, unsure what level is appropriate though. The only place I was somewhat unsure of is for the password.

This is now also mentioned in the commit body.

Comment (poettering) [reply src/firstboot/firstboot.c]: I think log_info() would still be good, to explain what goes on. Debug means it's hidden by default

File: meson.build [hunk 3]
```diff
@@ -982,6 +982,9 @@ if default_keymap == ''
 endif
 conf.set_quoted('SYSTEMD_DEFAULT_KEYMAP', default_keymap)
 
+default_timezone = get_option('default-timezone')
+conf.set_quoted('SYSTEMD_DEFAULT_TIMEZONE', default_timezone)
```
Comment (poettering) [inline meson.build]: so, there is quite an asymmetry here. SYSTEMD_DEFAULT_KEYMAP and SYSTEMD_DEFAULT_LOCALE both get applied in *absence* of any config. But this one here instead is about initializing the config with a default value, so it means in the end there *will* be a config.

That's quite a distinction, and I am not sure I like it.

I see no nice way to make SYSTEMD_DEFAULT_TIMEZONE work like the others admittedly though. After all, glibc never checks anything but /etc/localtime (well and $TZ, but that's nothing we can reasonably set, because it would disallow changingthe TZ later), and that's something we cannot really control ahead of time.

Hence maybe this is OK to add like this, but please rename this to something such as SYSTEMD_FIRSTBOOT_TIMEZONE or so, to indicate that this works differently from the others.

File: src/firstboot/firstboot.c [hunk 4]
```diff
@@ -675,13 +681,22 @@ static int prompt_hostname(int rfd) {
 
         for (;;) {
                 _cleanup_free_ char *h = NULL;
+                _cleanup_free_ char *fallback = NULL;
```
Comment (poettering) [inline src/firstboot/firstboot.c]: please merge these two lines

File: src/firstboot/firstboot.c [hunk 5]
```diff
@@ -675,13 +681,22 @@ static int prompt_hostname(int rfd) {
...
 
-                r = ask_string(&h, "%s Please enter hostname for new system (empty to skip): ", special_glyph(SPECIAL_GLYPH_TRIANGULAR_BULLET));
+                fallback = get_default_hostname();
```
Comment (poettering) [inline src/firstboot/firstboot.c]: lacks oom check.

but i really don't like this approach. we shouldn't write out redundant config files. i.e. if the compiled-in default is x, and this is a setting where we already set x anyway if we find no config, then let's *not* write out a config that sets it. The less config we write, the btter.

This matters actaully, so that on upgrades these settings follow our defaults, and do not remain "sticky" to whatever was current on first install. i.e. we want that /etc/ ideally contains only *user* selection, and not automatic choices – to the level this is possible.

File: src/basic/strv.c [hunk 6]
```diff
@@ -66,6 +66,36 @@ char* strv_find_startswith(char * const *l, const char *name) {
         return NULL;
 }
 
+char* strv_find_closest(char * const *l, const char *name) {
+        ssize_t best_distance = SSIZE_MAX;
+        char *best = NULL;
+
+        assert(l);
```
Comment (poettering) [inline src/basic/strv.c]: we generally treat a NULL strv like an empty strv. Hence please do not add an assert like this

File: src/basic/strv.c [hunk 7]
```diff
@@ -66,6 +66,36 @@ char* strv_find_startswith(char * const *l, const char *name) {
...
+
+        if (!name)
+                return NULL;
+
+        for (size_t i = 0; l[i]; i++) {
```
Comment (poettering) [inline src/basic/strv.c]: please use STRV_FOREACH()

File: src/basic/strv.h [hunk 8]
```diff
@@ -17,6 +17,7 @@ char* strv_find(char * const *l, const char *name) _pure_;
 char* strv_find_case(char * const *l, const char *name) _pure_;
 char* strv_find_prefix(char * const *l, const char *name) _pure_;
 char* strv_find_startswith(char * const *l, const char *name) _pure_;
+char* strv_find_closest(char * const *l, const char *name) _pure_;
```
Comment (poettering) [inline src/basic/strv.h]: maybe name this strv_find_closest_by_levenshtein()? this is otherwise so unclear

File: meson.build (hunk 3 above)
Comment (NekkoDroid) [reply meson.build]: I renamed it.

> But this one here instead is about initializing the config with a default value, so it means in the end there will be a config.

This is kind of done for all `sd-firstboot` stuff, this includes keymap and locale. If the default shouldn't be written there are 2 ways to prevent it:

- Don't `--prompt` the respective option
- Pass `-` to the respective prompt

And most of these prompts are very user specific and not something that really can be decided/should be changed by the distro, the only one that this doesn't apply to is the root shell.

File: src/firstboot/firstboot.c (hunk 5 above)
Comment (NekkoDroid) [reply src/firstboot/firstboot.c]: See my comment on the new config value for more detailed.

TLDR: All the other prompts are also written when prompted and selecting the default and there are ways to not have it write the config.

**Comment by NekkoDroid** [comment]
Now thinking about the original suggested improvements, maybe adding the `firstboot-timezone` option isn't too good of an idea if all get implemented at some point. Specifically suggesting defaults depending on previous things that where selected (e.g. suggesting `de-latin1` and `Europe/Berlin` as defaults when someone entered `de` as locale).

File: meson.build [hunk 9]
```diff
@@ -982,6 +982,9 @@ if default_keymap == ''
 endif
 conf.set_quoted('SYSTEMD_DEFAULT_KEYMAP', default_keymap)
 
+default_timezone = get_option('firstboot-timezone')
+conf.set_quoted('SYSTEMD_FIRSTBOOT_TIMEZONE', default_timezone)
```
Comment (yuwata) [inline meson.build]: I do not think we should provide such compiled default value. That's mostly useless, except for when e.g. systemd is built for a local company or so.

If you want to introduce a way to provide a default timezone. Please consider to use an environment variable or a kernel command line option.

**Review by yuwata — CHANGES_REQUESTED** [review]
Sorry for late review.

**Comment by yuwata** [comment]
The first two and the last commits are now waiting in #34235.

File: meson.build (hunk 9 above)
Comment (NekkoDroid) [reply meson.build]: Yea I wasn't sure about this. I kinda did it so the default can be changed, but it was more of a stopgap solution since the suggested plan was to have the default fallback (where this is used) be determined by the entered locale or something. So I just removed it and hardcoded `UTC` for now.

File: src/firstboot/firstboot.c [hunk 10]
```diff
@@ -592,7 +607,7 @@ static int prompt_timezone(int rfd) {
         print_welcome(rfd);
 
         r = prompt_loop("Please enter timezone name or number",
-                        zones, 30, timezone_is_valid_log_debug, &arg_timezone);
+                        zones, "UTC", 30, timezone_is_valid_log_debug, &arg_timezone);
```
Comment (NekkoDroid) [inline src/firstboot/firstboot.c]: I am not entirely happy with this. Would it make sense to instead of hardcoding the compile time values to query the current value and set based on that? so here `get_timezone` and what would the one for locale be?

This should then in theory default to the compile time values unless they have already been manually changed, in which case it can somewhat safely be assumed that that value is more likely to be wanted.

**Review by yuwata — COMMENTED** [review]
I like the second commit. Could you open a new PR with the second commit ?

**Comment by yuwata** [comment]
Please rebase.

**Comment by NekkoDroid** [comment]
A slight inconsistency is the `kernel-command-line` option. That is the only option that can only be set via the application cmdline and isn't read from any credential or prompt from a user. It seems really odd to have this here in general as the commandline needs to kinda be set up before even being able to boot a new system.

It also currently unconditionally creates the `/etc/kernel` directory even if not configured, resulting in an empty directory.

File: src/firstboot/firstboot.c [hunk 11]
```diff
@@ -323,15 +323,15 @@ static int prompt_locale(int rfd) {
 
                 print_welcome(rfd);
 
-                r = prompt_loop("Please enter system locale name or number",
+                r = prompt_loop("Please enter the new locale name or number",
```
Comment (yuwata) [inline src/firstboot/firstboot.c]: For locale, please keep 'system':
`Please enter the new system locale name or number`

File: src/firstboot/firstboot.c [hunk 12]
```diff
@@ -323,15 +323,15 @@ static int prompt_locale(int rfd) {
...
                                 locales, 60, is_valid, &arg_locale);
                 if (r < 0)
                         return r;
 
                 if (isempty(arg_locale))
                         return 0;
 
-                r = prompt_loop("Please enter system message locale name or number",
+                r = prompt_loop("Please enter the new message locale name or number",
```
Comment (yuwata) [inline src/firstboot/firstboot.c]: Also here.

File: src/firstboot/firstboot.c [hunk 13]
```diff
@@ -155,27 +156,29 @@ static int prompt_loop(const char *text, char **l, unsigned percentage, bool (*i
...
+                                        return r;
 
-                r = safe_atou(p, &u);
-                if (r >= 0) {
-                        if (u <= 0 || u > strv_length(l)) {
-                                log_error("Specified entry number out of range.");
+                                putchar('\n');
                                 continue;
-                        }
+                        };
```
Comment (yuwata) [inline src/firstboot/firstboot.c]: Not your fault, but unnecessary `;` ?

File: src/firstboot/firstboot.c [hunk 14]
```diff
@@ -540,14 +546,18 @@ static int prompt_timezone(int rfd) {
                 return 0;
         }
 
+        r = get_timezone(&zone);
+        if (r < 0)
+                return log_error_errno(r, "Cannot query timezone: %m");
```
Comment (yuwata) [inline src/firstboot/firstboot.c]: I think it is better to fall back to "UTC".

File: src/firstboot/firstboot.c [hunk 15]
```diff
@@ -145,14 +145,19 @@ static int prompt_loop(int rfd, const char *text, char **l, unsigned percentage,
                 _cleanup_free_ char *p = NULL;
                 unsigned u;
 
-                r = ask_string(&p, strv_isempty(l) ? "%s %s (empty to skip): "
-                                                   : "%s %s (empty to skip, \"list\" to list options): ",
-                               special_glyph(SPECIAL_GLYPH_TRIANGULAR_BULLET), text);
+                r = ask_string(&p, strv_isempty(l) ? "%s %s (empty to default to \"%s\", \"-\" to skip): "
+                                                   : "%s %s (empty to default to \"%s\", \"-\" to skip, \"list\" to list options): ",
```
Comment (yuwata) [inline src/firstboot/firstboot.c:149]: I'd like to keep the behavior for an empty string. How about the following?
- empty string -> skip
- "default" -> explicitly set default string
- "list" -> list options

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good to me.

File: src/firstboot/firstboot.c (hunk 15 above)
Comment (NekkoDroid) [reply src/firstboot/firstboot.c:149]: I was mostly following what was suggested in this comment https://github.com/systemd/systemd/pull/31154#issuecomment-1946056086 (Point 1 and 6).

I am not a fan of using the literal "default" to check for that, but I also don't know what really could be used for that other than how it currently is.

File: src/firstboot/firstboot.c (hunk 14 above)
Comment (NekkoDroid) [reply src/firstboot/firstboot.c]: I changed it to `UTC` for now.

I think it probably would make more sense to default these values to the values of the live system. Usually you'd likely change the keymap and locale on an install medium before invoking something like `sd-firstboot` on what you are about to install to.

Other than the current timezone I didn't see any function to get the current locale and keymap.

File: src/firstboot/firstboot.c (hunk 15 above)
Comment (yuwata) [reply src/firstboot/firstboot.c:149]: Not a strong opinion, but I'd like to keep the behavior for an empty string, as explicitly setting default value is typically not necessary. IIRC, @poettering also commented similar in the above (not searched).

"-" may be used for the literal for applying default explicitly, if "default" is not good.

Comment (NekkoDroid) [reply src/firstboot/firstboot.c:149]: > IIRC, @poettering also commented similar in the above (not searched).

He did, just he didn't follow up on my clarification.

I'd say that that commit is still something that needs more work and discussion, as mentioned above the defaults/fallbacks used currently could be better than just the compile time default.

Comment (yuwata) [reply src/firstboot/firstboot.c:149]: @poettering Could you comment on this?

There are two reasons I want to swap an empty string and "-":
- (already commented in the above, but again) explicitly setting default value is typically not necessary
- We usually handle an empty string as doing compiled time default, especially in unit files or .conf files. So, this conflicts with usual behaviors of other components.

> > IIRC, @poettering also commented similar in the above (not searched).
> 
> He did, just he didn't follow up on my clarification.

Maybe https://github.com/systemd/systemd/pull/33226#discussion_r1648902754 ? But I cannot follow your reply.

> I'd say that that commit is still something that needs more work and discussion, as mentioned above the defaults/fallbacks used currently could be better than just the compile time default.

Maybe. But, even so, I'd like to keep empty to do nothing.

Comment (NekkoDroid) [reply src/firstboot/firstboot.c:149]: > Maybe https://github.com/systemd/systemd/pull/33226#discussion_r1648902754 ? But I cannot follow your reply.

I was talking about this comment https://github.com/systemd/systemd/pull/33226#discussion_r1649455325 (forgot I marked that discussion as resolved)

My arguments were:

1. These prompts need to be opted in to with either `--prompt` or the respective `--prompt-*` option so the user already opted into setting some value for these (although with the `sd-firstboot.service` change this argument is debatable)
1. These few configuration files writen in this service aren't really something that shouldn't exist on a system (this already caused us to copy the default file from `/usr/share/factory/` https://github.com/systemd/systemd/pull/31154#issuecomment-1946174780)
1. and these are highly system specific and not distribution specific and changes to the values from a distribution side could be surprising, so "using distro default" in this case should most likely be opt-in.

**Comment by NekkoDroid** [comment]
I'd like to get the first 4 commits into the release since those seem rather uncontroversial, would that be possible?

File: src/firstboot/firstboot.c (hunk 15 above)
Comment (NekkoDroid) [reply src/firstboot/firstboot.c:149]: > - empty string -> skip
> - "default" -> explicitly set default string
> - "list" -> list options

@yuwata I just had the thought: what about instead of "default" using `<tab>`? I feel that makes a lot of sense since it's already use for completion in shells, so it would also follow that it would work here.

**Comment by yuwata** [comment]
> I'd like to get the first 4 commits into the release since those seem rather uncontroversial, would that be possible?

I agree. Please open a new PR with the first 4 commits. Then, I will approve it soon.

File: src/firstboot/firstboot.c (hunk 15 above)
Comment (yuwata) [reply src/firstboot/firstboot.c:149]: > > * empty string -> skip
> > * "default" -> explicitly set default string
> > * "list" -> list options
> 
> @yuwata I just had the thought: what about instead of "default" using `<tab>`? I feel that makes a lot of sense since it's already use for completion in shells, so it would also follow that it would work here.

Using tab sounds OK to me.

Comment (NekkoDroid) [reply src/firstboot/firstboot.c:149]: I was just looking into this. Either I was missing some existing APIs or it would need to be implemented (as just checking for a returned tab is rather ugly on the user side), at which point it might just make more sense to skip this entire "default selection" stuff and implement full tab completion.

**Comment by poettering** [comment]
needs rebase

**Comment by poettering** [comment]
so I'd be fine if there was a way to explicitly write the default value to disk, but i think we shouldn't do that by default, but on request. if that request is a cmdline switch, or a special key you hit on input (such as TAB), or even a special string you enter for it is all fine be me. all I care about is that if people just click through the wizard quickly without entering anything we do not end up with these files by default.

**Comment by poettering** [comment]
(I think tab completion would be kinda awesome)
